src/common/src/config.rs

```diff
@@ -888,6 +888,12 @@ pub struct StreamingDeveloperConfig {
     /// The max heap size of dirty groups of `HashAggExecutor`.
     #[serde(default = "default::developer::stream_hash_agg_max_dirty_groups_heap_size")]
     pub hash_agg_max_dirty_groups_heap_size: usize,
+
+    #[serde(default = "default::developer::stream_disable_arrangement_backfill")]
+    /// Disable arrangement backfill
+    /// If true, the arrangement backfill will be disabled,
+    /// even if session variable set.
+    pub disable_arrangement_backfill: bool,
```

Comment:
So if user can control whether to use arrangement backfill by `streaming_use_arrangement_backfill`, what's the point of this config item?

BTW, I'd suggest to use a _positive_ verb (`enable`) instead of a negative one, if this is really needed. Otherwise it can be easily confused.

Reply:
> streaming_use_arrangement_backfill

This is a session variable, so we need to user to set it per session to disable it.
With `disable_arrangement_backfill`, the user disable it for the cluster, no need to set per session.

Reply:
It seems that if the purpose is "no need to set per session", the cluster config item should be used as a "fallback value", and should be sth like `default_use_arrangement_backfill`. I don't see the point if the cluster config is used to disable it ignoring session config.🤔

Reply:
I guess we are introducing some hierarchical configuration, so that for a given config item, the cluster level config value serves as a default fallback value, and session level config value can override it, and mv level config value can further override the session level one.

Reply:
The problem is that session config is always "set". Because it will always have a default value if none is set.

Reply:
May involve @yuhao-su to give some insight.

Reply:
It can be solved if we change `use_arrangement_backfill` to have 3 values:
```
true | false | unset
```

Reply:
Yes, I guess to introduce a hierarchical configuration I mentioned above, every config item should have a value called `unset`. I'm not sure if Yuhao is doing this way but I guess this can be the most reasonable design.

BTW it's not a strong opinion here, since we haven't got the new config system finished.


> I'd suggest to use a _positive_ verb (`enable`) instead of a negative one, if this is really needed. Otherwise it can be easily confused.

While this is a kinda strong suggestion cuz I did meet similar problem before. People can easily confused by "true meaning disable, false meaning enable".

Reply:
> While this is a kinda strong suggestion cuz I did meet similar problem before. People can easily confused by "true meaning disable, false meaning enable".

Fixed https://github.com/risingwavelabs/risingwave/pull/15692/commits/b8e4514ae8824b7beb7163cf1604640af2df9bca

Reply:
In the configuration I'm refactoring now, a session variable default value can be set system-wide. 
Usually, a session variable should override the system-wild variable. If the case is to have a special switch to force all arrangement backfill disabled (regardless of any session variable), then I agree you can have this in the developer config.

Reply:
> In the configuration I'm refactoring now, a session variable default value can be set system-wide.
Usually, a session variable should override the system-wild variable. If the case is to have a special switch to force all arrangement backfill disabled (regardless of any session variable), then I agree you can have this in the developer config.

The goal is to just have a way to disable arrangement backfill, with or without session variable.
Eventually, once your refactoring is complete, I would like to have hierachical config, as per your refactor.
That will introduce a breaking change to the behaviour of the `enable_arrangement_backfill` eventually.

I think that's acceptable, if users set `enable_arrangement_backfill` in the cluster config, it's typically due to intervention from rw developers, since we know it's some problem due to arrangement backfill, and are likely to disable it for the entire cluster, without wanting to override with session variable for specific queries.

So down the road, once your refactor is complete, we can make it more flexible, where we let session variables override as well.

Note that it won't cause breaking change in metadata, because `cluster` config is only used on startup. It is not a system config. Only breaking change in its behaviour. i.e. if user relies on the cluster wide `enable_arrangement_backfill` config being set to `false`, always ignore session variable setting. But that's unlikely IMO.